sycl/include/sycl/ext/oneapi/experimental/chunk.hpp

```diff
@@ -122,30 +129,49 @@ template <size_t ChunkSize, typename ParentGroup> class chunk {
   }
 
 protected:
+  static constexpr size_t chunk_size = ChunkSize;
+
 #if defined(__SYCL_DEVICE_ONLY__) && defined(__NVPTX__)
   sub_group_mask Mask;
-#endif
 
-#if defined(__SYCL_DEVICE_ONLY__) && defined(__NVPTX__)
   chunk(ext::oneapi::sub_group_mask mask) : Mask(mask) {}
+
+  ext::oneapi::sub_group_mask getMask() const { return Mask; }
 #else
   chunk() {}
+
+#ifdef __SYCL_DEVICE_ONLY__
+  ext::oneapi::sub_group_mask getMask() const {
+    ext::oneapi::sub_group_mask::BitsType MaskBits{0};
+    MaskBits = ~MaskBits;
+    MaskBits >>= ext::oneapi::sub_group_mask::max_bits - ChunkSize;
+    MaskBits <<=
+        ext::oneapi::sub_group_mask::max_bits -
+        (((__spirv_SubgroupLocalInvocationId() / ChunkSize) + 1) * ChunkSize);
+    return sycl::detail::Builder::createSubGroupMask<
+        ext::oneapi::sub_group_mask>(MaskBits, __spirv_SubgroupMaxSize());
+  }
+#endif
```

Comment:
I'm having trouble following the logic here, so I think a comment would help.

Is this supposed to represent the mask within the entire sub-group always, or just the mask within the current chunk?  I'm trying to work out why it's okay to unconditionally use `LocalInvocationId` and `ChunkSize` here, but we have to account for chunks-of-chunks with a dedicated branch above.

Reply:
It's the active mask of the current chunk. That is, if the chunk size is 2 and the sub-group size is 8, it's one of
11000000 (sub-group LID 0 and 1)
00110000 (sub-group LID 2 and 3)
00001100 (sub-group LID 4 and 5)
00000011 (sub-group LID 6 and 7)

Let's say for example we have LID 5, then the steps would be:
145: MaskBits = 00000000
146: MaskBits = 11111111 (~00000000)
147: MaskBits = 00000011 (11111111 >> (8 - 2))
148: MaskBits = 00001100 (00000011 << (8 - (floor(5 / 2) + 1) * 2))

Reply:
But why don't you need to account differently for chunks of chunks vs chunks of entire sub-groups?

Is it just because everything is guaranteed to be power-of-two aligned?  So splitting a sub-group into chunks of 4 and _then_ into chunks of 2 gives us the same mask as splitting into chunks of 2 directly?

I think that makes sense, but I wanted to make sure I understood it before approving.

Reply:
That's exactly right! To be fair, I was half way into passing masks around as members when I realized the masks were really independent of how many layers of chunks were above. 😄